api/v1/storageprofile_types.go

```diff
@@ -24,25 +24,25 @@ import (
 // NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
 
 // StorageProfileSpec defines the desired state of StorageProfile
-// +kubebuilder:validation:XValidation:rule="!has(oldSelf.deviceClass) || has(self.deviceClass)", message="deviceClass is required once set"
-// +kubebuilder:validation:XValidation:rule="!has(oldSelf.sharedFilesystemConfiguration) || has(self.sharedFilesystemConfiguration)", message="sharedFilesystemConfiguration is required once set"
-// +kubebuilder:validation:XValidation:rule="!has(oldSelf.blockPoolConfiguration) || has(self.blockPoolConfiguration)", message="blockPoolConfiguration is required once set"
+// +kubebuilder:validation:XValidation:rule="has(oldSelf.deviceClass) && has(self.deviceClass)", message="deviceClass is required"
+// +kubebuilder:validation:XValidation:rule="has(oldSelf.sharedFilesystemConfiguration) && has(self.sharedFilesystemConfiguration)", message="sharedFilesystemConfiguration is required"
+// +kubebuilder:validation:XValidation:rule="has(oldSelf.blockPoolConfiguration) && has(self.blockPoolConfiguration)", message="blockPoolConfiguration is required"
 type StorageProfileSpec struct {
 	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
 	// Important: Run "make" to regenerate code after modifying this file
 
-	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Required
 	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="DeviceClass is immutable"
 	// +kubebuilder:validation:MaxLength=512
 	// DeviceClass is the deviceclass name.
 	DeviceClass string `json:"deviceClass"`
 
-	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Required
 	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="SharedFilesystemConfiguration is immutable"
 	// configurations to use for cephfilesystem.
 	SharedFilesystemConfiguration SharedFilesystemConfigurationSpec `json:"sharedFilesystemConfiguration,omitempty"`
 
-	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Required
 	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="BlockPoolConfiguration is immutable"
```

Comment:
i did the below changes and am getting as below results

```go
 // StorageProfileSpec defines the desired state of StorageProfile
-// +kubebuilder:validation:XValidation:rule="!has(oldSelf.deviceClass) || has(self.deviceClass)", message="deviceClass is required once set"
-// +kubebuilder:validation:XValidation:rule="!has(oldSelf.sharedFilesystemConfiguration) || has(self.sharedFilesystemConfiguration)", message="sharedFilesystemConfiguration is required once set"
-// +kubebuilder:validation:XValidation:rule="!has(oldSelf.blockPoolConfiguration) || has(self.blockPoolConfiguration)", message="blockPoolConfiguration is required once set"
 type StorageProfileSpec struct {
        // INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
        // Important: Run "make" to regenerate code after modifying this file
 
-       // +kubebuilder:validation:Optional
-       // +kubebuilder:validation:XValidation:rule="self == oldSelf",message="DeviceClass is immutable"
       // +kubebuilder:validation:MaxLength=512
+       // +kubebuilder:validation:Required
        // DeviceClass is the deviceclass name.
        DeviceClass string `json:"deviceClass"`
 
        // +kubebuilder:validation:Optional
-       // +kubebuilder:validation:XValidation:rule="self == oldSelf",message="SharedFilesystemConfiguration is immutable"
        // configurations to use for cephfilesystem.
        SharedFilesystemConfiguration SharedFilesystemConfigurationSpec `json:"sharedFilesystemConfiguration,omitempty"`
 
        // +kubebuilder:validation:Optional
-       // +kubebuilder:validation:XValidation:rule="self == oldSelf",message="BlockPoolConfiguration is immutable"
        // configurations to use for  profile specific blockpool.
        BlockPoolConfiguration BlockPoolConfigurationSpec `json:"blockPoolConfiguration,omitempty"`
 }
@@ -67,7 +60,9 @@ type StorageProfile struct {
        metav1.TypeMeta   `json:",inline"`
        metav1.ObjectMeta `json:"metadata,omitempty"`
 
-       Spec   StorageProfileSpec   `json:"spec,omitempty"`
+       // +kubebuilder:validation:Required
+       // +kubebuilder:validation:XValidation:rule="self == oldSelf",message="spec is immutable"
+       Spec   StorageProfileSpec   `json:"spec"`
        Status StorageProfileStatus `json:"status,omitempty"`
 }
```

* Not able to create the SP without Spec or DeviceClass
* Not able to delete or update the DeviceClass once SP is created
* Not able to Update or Delete the sharedFilesystemConfiguration/blockPoolConfiguration once SP is created

Let me know if this is what you are looking for

Reply:
@Madhu-1 as mentioned, I am making changes to the code! Re-read [comment](https://github.com/red-hat-storage/ocs-operator/pull/2231#discussion_r1371899791)

Reply:
@Madhu-1 for the fields `sharedFilesystemConfiguration`and `blockPoolConfiguration` your suggestion allows transitioning from unset to set, or from set to unset, which is not what's required.

Reply:
@raaizik when i tested it i was not able to i was not able to add the sharedFilesystemConfiguration when it was not present , see below output





```yaml
[🎩︎]mrajanna@li-2cfbef4c-22d9-11b2-a85c-a3e4a93c405f ocs-operator $]kubectl get storageprofile -oyaml
apiVersion: v1
items:
- apiVersion: ocs.openshift.io/v1
  kind: StorageProfile
  metadata:
    creationTimestamp: "2023-10-30T09:18:09Z"
    generation: 1
    labels:
      app.kubernetes.io/created-by: ocs-operator
      app.kubernetes.io/instance: storageprofile-sample
      app.kubernetes.io/managed-by: kustomize
      app.kubernetes.io/name: storageprofile
      app.kubernetes.io/part-of: ocs-operator
    name: storageprofile-sample
    namespace: default
    resourceVersion: "19428"
    uid: cfd7f5cb-6e56-4dbf-8bed-66cd93982278
  spec:
    deviceClass: test
kind: List
metadata:
  resourceVersion: ""
```

When i tried to edit the SP to add the sharedFilesystemConfiguration it failed saying the spec is immutable

```yaml
# Please edit the object below. Lines beginning with a '#' will be ignored,
# and an empty file will abort the edit. If an error occurs while saving this file will be
# reopened with the relevant failures.
#
# storageprofiles.ocs.openshift.io "storageprofile-sample" was not valid:
# * spec: Invalid value: "object": spec is immutable
#
apiVersion: ocs.openshift.io/v1
kind: StorageProfile
metadata:
  creationTimestamp: "2023-10-30T09:18:09Z"
  generation: 1
  labels:
    app.kubernetes.io/created-by: ocs-operator
    app.kubernetes.io/instance: storageprofile-sample
    app.kubernetes.io/managed-by: kustomize
    app.kubernetes.io/name: storageprofile
    app.kubernetes.io/part-of: ocs-operator
  name: storageprofile-sample
  namespace: default
  resourceVersion: "19428"
  uid: cfd7f5cb-6e56-4dbf-8bed-66cd93982278
spec:
  deviceClass: test
  sharedFilesystemConfiguration:
    parameters:
      test: test
```


Am i testing something wrong or if am missing anything please let me know :)

Reply:
@Madhu-1 You're missing rules around `parameters` sub-fields of both `sharedFilesystemConfigurationand` and `blockPoolConfiguration`, that allows for (un)sets post creation. Not critical, my new suggestion uses a part of yours, take a look.